planner/core/rule_join_reorder.go

```diff
@@ -31,25 +31,43 @@ import (
 // construct a new join order based on a reorder algorithm.
 //
 // For example: "InnerJoin(InnerJoin(a, b), LeftJoin(c, d))"
-// results in a join group {a, b, LeftJoin(c, d)}.
-func extractJoinGroup(p LogicalPlan) (group []LogicalPlan, eqEdges []*expression.ScalarFunction, otherConds []expression.Expression) {
+// results in a join group {a, b, c, d}.
+func extractJoinGroup(p LogicalPlan) (group []LogicalPlan, eqEdges []*expression.ScalarFunction,
+	otherConds []expression.Expression, joinTypes []JoinType) {
 	join, isJoin := p.(*LogicalJoin)
-	if !isJoin || join.preferJoinType > uint(0) || join.JoinType != InnerJoin || join.StraightJoin {
-		return []LogicalPlan{p}, nil, nil
+	if !isJoin || join.preferJoinType > uint(0) || join.StraightJoin ||
+		(join.JoinType != InnerJoin && join.JoinType != LeftOuterJoin && join.JoinType != RightOuterJoin) ||
+		((join.JoinType == LeftOuterJoin || join.JoinType == RightOuterJoin) && join.EqualConditions == nil) {
+		return []LogicalPlan{p}, nil, nil, nil
+	}
+	if join.JoinType != RightOuterJoin {
+		lhsGroup, lhsEqualConds, lhsOtherConds, lhsJoinTypes := extractJoinGroup(join.children[0])
+		group = append(group, lhsGroup...)
+		eqEdges = append(eqEdges, lhsEqualConds...)
+		otherConds = append(otherConds, lhsOtherConds...)
+		joinTypes = append(joinTypes, lhsJoinTypes...)
+	} else {
+		group = append(group, join.children[0])
 	}
 
-	lhsGroup, lhsEqualConds, lhsOtherConds := extractJoinGroup(join.children[0])
-	rhsGroup, rhsEqualConds, rhsOtherConds := extractJoinGroup(join.children[1])
+	if join.JoinType != LeftOuterJoin {
+		rhsGroup, rhsEqualConds, rhsOtherConds, rhsJoinTypes := extractJoinGroup(join.children[1])
+		group = append(group, rhsGroup...)
+		eqEdges = append(eqEdges, rhsEqualConds...)
+		otherConds = append(otherConds, rhsOtherConds...)
+		joinTypes = append(joinTypes, rhsJoinTypes...)
+	} else {
+		group = append(group, join.children[1])
+	}
 
-	group = append(group, lhsGroup...)
-	group = append(group, rhsGroup...)
 	eqEdges = append(eqEdges, join.EqualConditions...)
-	eqEdges = append(eqEdges, lhsEqualConds...)
-	eqEdges = append(eqEdges, rhsEqualConds...)
 	otherConds = append(otherConds, join.OtherConditions...)
-	otherConds = append(otherConds, lhsOtherConds...)
-	otherConds = append(otherConds, rhsOtherConds...)
-	return group, eqEdges, otherConds
+	otherConds = append(otherConds, join.LeftConditions...)
+	otherConds = append(otherConds, join.RightConditions...)
+	for range join.EqualConditions {
+		joinTypes = append(joinTypes, join.JoinType)
+	}
```

Comment:
What is the purpose of the `for` loop here?

Reply:
The record joinType is used to func checkConnectionAndMakeJoin

Reply:
Why not use
```
joinTypes = append(joinTypes, join.JoinType)
```
rather than
```
for range join.EqualConditions {
		joinTypes = append(joinTypes, join.JoinType)
	}
```

Reply:
Why not use 
```
joinTypes = append(joinTypes, join.JoinType)
```
rather than
```
for range join.EqualConditions {
		joinTypes = append(joinTypes, join.JoinType)
}
```
.

If there are conditions like `t1 join t2 on t1.a = t2.a and t1.b = t2.b`, there are two equal conditions. But we only need one joinType?